c10/core/TensorTypeSet.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <c10/core/TensorTypeId.h>
+#include <c10/util/llvmMathExtras.h>
+#include <c10/util/Exception.h>
+#include <ostream>
+
+namespace c10 {
+
+// A representation of a set of TensorTypeIds.  A tensor may
+// have multiple tensor type ids, e.g., a Variable tensor
+// can also be a CPU tensor; the TensorTypeSet specifies what
+// type ids apply.  The internal representation is as a 64-bit
+// bit set (this means only 64 tensor type ids are supported).
+//
+// At the moment, there are no nontrivial uses of this set; tensors
+// are always singletons.  In the near future, this set will represent
+// variable? + tensor type id.  In the far future, it will be
+// variable? + profiling? + tracing? + lazy? + tensor type id.
+//
+// An undefined tensor is one with an empty tensor type set.
+class TensorTypeSet {
+public:
+  TensorTypeSet() {}
+  explicit TensorTypeSet(TensorTypeId t)
+    : repr_(t == TensorTypeId::UndefinedTensorId
+              ? 0
+              : 1ULL << (static_cast<uint8_t>(t) - 1)) {}
+  // Test if a TensorTypeId is in the set
+  bool has(TensorTypeId t) const {
+    TORCH_INTERNAL_ASSERT(t != TensorTypeId::UndefinedTensorId);
+    return static_cast<bool>(repr_ & TensorTypeSet(t).repr_);
+  }
+  // Perform set union
+  TensorTypeSet operator|(TensorTypeSet other) const {
```

Comment:
is there precedence for this kind of design?  I thought given you have functions called add, remove, etc. this would be called union?  Or, everything would be Bit based and you'd just expose those operators.

Reply:
Python sets overload bitwise operators to mean union, so I'm taking a page from their API book. A semantic model of TensorTypeSet as a bit set is pretty accurate, and I think these overloads help.

Reply:
ya, I think the only weirdness to me is that you also don't provide `union` in addition to `operator|` (this is minor).  Python always gives you names, and also overloads the operators in places they make sense.  It feels like a mixed metaphor if sometimes we have names, and sometimes we have overloaded operators.

Reply:
Going to fix your comment, I realized another reason I didn't add a union method: union is a reserved keyword in C++, so I can't :)

Reply:
that seems like a good reason :).